File: Hard-Array/Largestsum0.java
import java.util.HashMap;
import java.util.Scanner;

class Largestsum0
{
    /**
     * This class contains a method to find the length of the largest subarray 
     * with sum equal to 0 in a given integer array.
     */
    public static int maxLength(int arr[])
    {
        /**
         * Uses a HashMap to store prefix sums and their first occurrence index.
         * For each element, maintain the running sum.
         * If the sum is zero, the subarray from 0 to i has sum 0.
         * If the sum has been seen before, the subarray between previous index+1 and i has sum 0.
         * Keep track of the maximum such length.
         */
        HashMap<Integer, Integer> map = new java.util.HashMap<>();
        int maxLen = 0;
        int sum = 0;
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
            if (sum == 0) {
                // Subarray from 0 to i has sum 0
                maxLen = i + 1;
            }
            if (map.containsKey(sum)) {
                // Subarray from map.get(sum)+1 to i has sum 0
                int prevIndex = map.get(sum);
                maxLen = Math.max(maxLen, i - prevIndex);
            } else {
                // Store first occurrence of this sum
                map.put(sum, i);
            }
        }
        return maxLen;
    }
    public static void main(String[] args)
    {
        Scanner sc=new Scanner(System.in);
        // Prompt user to enter the size of the array
        System.out.println("Enter the size of array");
        int n=sc.nextInt();
        int arr[]=new int[n];
        // Prompt user to enter the elements of the array
        System.out.println("Enter the elements of array");
        for(int i=0;i<n;i++)
        {
            arr[i]=sc.nextInt();    
        }
        // Call maxLength method to find the largest subarray with sum zero
        int result=maxLength(arr);
        // Print the result
        System.out.println("The length of largest subarray with sum 0 is: "+result);
    }
}

File: Hard-Array/Largestsum0.py
# WAP to find the Given an array containing both positive and negative integers, we have to find the length of the longest subarray with the sum of all elements equal to zero.

# brute froce approach
def longest_zero_sum_bruteforce(arr):
    # Get the length of the array
    n = len(arr)
    
    # Variable to store the maximum length of subarray with sum = 0
    max_length = 0

    # First loop: choose the starting index of subarray
    for i in range(n):
        # Second loop: choose the ending index of subarray
        for j in range(i, n):
            # Initialize sum for the current subarray
            sub_sum = 0
            # Third loop: calculate sum of elements from index i to j
            for k in range(i, j + 1):
                sub_sum += arr[k]

            # Check if the current subarray sum is zero
            if sub_sum == 0:
                # Update max_length if this subarray is longer
                max_length = max(max_length, j - i + 1)

    # Return the length of the longest subarray with sum = 0
    return max_length

# better apporach
def longest_zero_sum_better(arr):
    n = len(arr)          # Total number of elements in array
    max_length = 0        # Stores the maximum length of subarray with sum = 0

    # Outer loop → selects the starting index of subarray
    for i in range(n):
        sub_sum = 0       # Reset sum for each new starting point

        # Inner loop → selects the ending index of subarray
        for j in range(i, n):
            sub_sum += arr[j]   # Add current element to subarray sum

            # If sum becomes 0 → valid subarray found
            if sub_sum == 0:
                # Update max_length if current subarray is longer
                max_length = max(max_length, j - i + 1)

    return max_length     # Return the maximum length found


nums=int(input("Enter the number of the elements :-"))
arr=[]
print("Enter the elements :-")
for i in range(nums):
    i=int(input())
    arr.append(arr[i])
result=longest_zero_sum_better(arr)
print("Longest length of the Array with sum zero = ",result)


    